File: Code/0509.Fibonacci Number/Fibonacci Number_3.py
class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        dp = [0, 1]
        for i in range(2, n+1):
            dp[0], dp[1] = dp[1], dp[0] + dp[1]     # 相较于第二种解法而言，节约了空间成本，不需要储存整个斐波那契数列
        return dp[1]

File: Code/0509.斐波那契数/Fibonacci Number_3.java
class Solution {
    public int fib(int n) {
        if(n <= 1){
            return n;
        }
        int[] dp = new int[2];
        // 动态数组初始化
        dp[0] = 0;
        dp[1] = 1;
        // 动态传递
        int temp;
        for(int i = 2;i <= n; i++){
            temp = dp[1];
            dp[1] = dp[0] + dp[1];
            dp[0] = temp;
            // dp[i] = dp[i-1] + dp[i-2];
        }
        return dp[1];
    }
}